core/domain/story_services.py

```diff
@@ -967,3 +968,67 @@ def record_completed_node_in_story_context(
         progress_model.completed_node_ids.append(node_id)
         progress_model.update_timestamps()
         progress_model.put()
+
+
+def get_chapter_notification_dicts() -> Dict[str, List[Dict[
+    str, Sequence[str]]]]:
+    """Returns a dict of behind schedule and upcoming chapters."""
+    topic_models = topic_fetchers.get_all_topics()
+    overdue_stories_dicts = []
+    upcoming_stories_dicts = []
+    for topic_model in topic_models:
+        topic_rights = topic_fetchers.get_topic_rights(topic_model.id)
+        if topic_rights.topic_is_published:
+            canonical_story_ids = [story.story_id
+                for story in topic_model.canonical_story_references]
+            canonical_story_models = story_models.StoryModel.get_multi(
+                canonical_story_ids)
+
+            for story_model in canonical_story_models:
+                story = (
+                    story_fetchers.get_story_from_model(story_model) if
+                    story_model else None)
+                if story is None:
+                    continue
+                overdue_chapters = []
+                upcoming_chapters = []
+                for node in story.story_contents.nodes:
+                    if node.planned_publication_date is not None and ((
+                        node.planned_publication_date -
+                        datetime.datetime.today()).days < 14 and
+                        node.planned_publication_date >
+                        datetime.datetime.today() and
+                        node.status != 'Published'):
+                        upcoming_chapters.append(node.title)
+
+                    if node.planned_publication_date is not None and (
+                        node.planned_publication_date <
+                        datetime.datetime.today()
+                        and node.status != 'Published'):
+                        overdue_chapters.append(node.title)
+
+                if len(overdue_chapters):
+                    story_link = (
+                        'https://www.oppia.org/story_editor/%s' % story.id)
```

Comment:
Do we hardcode this elsewhere (e.g. in other emails)? I thought we usually use a constant in feconf or similar to populate the first part of the URL.

In any case, you should only pass the strictly-necessary info that's related to this part of the domain logic -- so, in this case, the story ID should be sufficient. Ditto below.

Reply:
Done